hi-ml-azure/tests/testazure/test_azure_util.py

```diff
@@ -539,100 +613,382 @@ def _get_experiment_runs() -> List[MockRun]:
+
+
+@pytest.mark.parametrize("dummy_env_vars, expect_file_downloaded", [({}, True), ({util.ENV_LOCAL_RANK: "1"}, False)])
+@patch("azureml.core.Run", MockRun)
+def test_download_run_file(tmp_path: Path, dummy_env_vars: Dict[str, str], expect_file_downloaded: bool) -> None:
+    dummy_filename = "filetodownload.txt"
+    expected_file_path = tmp_path / dummy_filename
+
+    # mock the method 'download_file' on the AML Run class and assert it gets called with the expected params
+    mock_run = MockRun(run_id="id123")
+    mock_run.download_file = MagicMock(return_value=None)  # type: ignore
+    mock_run.download_file.side_effect = _mock_download_file
+
+    with mock.patch.dict(os.environ, dummy_env_vars):
+        util.download_run_file(mock_run, dummy_filename, expected_file_path)
+
+        if expect_file_downloaded:
+            mock_run.download_file.assert_called_with(dummy_filename, output_file_path=expected_file_path,
+                                                      _validate_checksum=False)
+            assert expected_file_path.exists()
+        else:
+            assert not expected_file_path.exists()
+
+
+def test_download_run_file_remote(tmp_path: Path) -> None:
+    # This test will create a Run in your workspace (using only local compute)
+    ws = DEFAULT_WORKSPACE.workspace
+    experiment = Experiment(ws, AML_TESTS_EXPERIMENT)
+    config = ScriptRunConfig(
+        source_directory=".",
+        command=["cd ."],  # command that does nothing
+        compute_target="local"
+    )
+    run = experiment.submit(config)
+
+    file_to_upload = tmp_path / "dummy_file.txt"
+    file_contents = "Hello world"
+    file_to_upload.write_text(file_contents)
+
+    # This should store the file in outputs
+    run.upload_file("dummy_file", str(file_to_upload))
+
+    output_file_path = tmp_path / "downloaded_file.txt"
+    assert not output_file_path.exists()
+
+    start_time = time.perf_counter()
+    util.download_run_file(run, "dummy_file", output_file_path)
+    end_time = time.perf_counter()
+    time_dont_validate_checksum = end_time - start_time
+
+    assert output_file_path.exists()
+    assert output_file_path.read_text() == file_contents
+
+    # Now delete the file and try again with _validate_checksum == True
+    output_file_path.unlink()
+    assert not output_file_path.exists()
+    start_time = time.perf_counter()
+    util.download_run_file(run, "dummy_file", output_file_path, validate_checksum=True)
+    end_time = time.perf_counter()
+    time_validate_checksum = end_time - start_time
+
+    assert output_file_path.exists()
+    assert output_file_path.read_text() == file_contents
+
+    logging.info(f"Time to download file without checksum: {time_dont_validate_checksum} vs time with"
+                 f"validation {time_validate_checksum}.")
+
+
+def test_download_run_file_during_run(tmp_path: Path) -> None:
+    # This test will create a Run in your workspace (using only local compute)
```

Comment:
As mentioned above, I don't get this test. The script uses the normal API to download files? 
If we wanted to test downloading during an AML run, I would suggest creating a local run first (as you are doing in the test above), and then submit a script to real AML that uses our custom download code, and NOT supply a workspace name. That script should download a file from that previous run, and check its contents, all in AML.

Reply:
To your first point - that was a mistake. It was supposed to call our own download_files 

To your second point - checking it's contents in AML, meaning that the AML run should fail if the contents are incorrect? What then would this test check for - whether that run succeeds?

Reply:
Replaced with Jonathan's test script pattern